### ISSUE TITLE
3DS normals

### ISSUE THREAD

**Issue by agerrius** [assimp/assimp#1086]
When importing a 3DS file the normals are calculated (obviously since the 3DS format doesn't store them), but I have the impression that the 3DS loader always smooths the normals, also when no smoothing group is assigned to the face.

Is this by design or am I missing some setting somewhere?

**Comment by smalcom** [comment]
May be in import flags you provide smoothing normals.

**Comment by agerrius** [comment]
I have tried with the generate smooth normals flag, the generate normals flag and also with no post-processing flag. All give the same result, so my impression is that the post-processing is not the issue.

When I look at the source code of the 3DS importer I think the function to generate the normals is called as part of the import (so before post-processing) and I think it doesn't include the smoothing group. But I need to study the code in more detail to be sure. That's why I asked here if it is by design.

**Comment by kimkulling** [comment]
As far as I understand the model formats these normals came from the smoohting groups defined in the 3DS-model. So when having these groups defined you will get smooth normals by design.

**Comment by agerrius** [comment]
Yes, but the problem is that also objects without a smoothing group get the smooth normals. I have attached a test object with two cylinders, one has a smoothing group, the other not. But when reading with Assimp they both look the same.

I'll continue to explore the Assimp code to see if I can understand what is going on.

[smooth.zip](https://github.com/assimp/assimp/files/617203/smooth.zip)

**Comment by kimkulling** [comment]
You are right. At the moment this will be done for all faces, which are part of the several submeshes( see 3DSLoader.cpp, Line 196:         ComputeNormalsWithSmoothingsGroups<D3DS::Face>(mesh);
)

**Comment by agerrius** [comment]
Yes, that was how I interpreted the code as well. I'll see if I can come up with a solution where the smoothing groups are taken into account correctly.

**Comment by kimkulling** [comment]
That would be really great. Maybe with a smla  unittest 👍 ?

:-D

**Comment by ov** [comment]
Hi guys, I've just meet the same issue as @agerrius and traced it down to `SGSpatialSort::FindPositions()`. There is a code at the very bottom of this method, that checks if smoothing group is zero and treats it as a smooth vertex/edge:

```c++
// if the given smoothing group is 0, we'll return all surrounding vertices
if (!pSG)
{
    ...
}
```

This gives you a smooth mesh if it misses a smooth groups information chunk in 3DS.  However, if you load such mesh in 3D Studio Max - you get all the edges hard:

![image](https://cloud.githubusercontent.com/assets/151412/26569590/9af64686-455f-11e7-8d25-2059e7c30919.png)

The solution seems to be to remove that check and treat zero smoothing group as "hard", but before submitting a patch I'd like to double-check if this is done intentionally. I see the same code is being used at ASE import, too, so maybe it needs to work a different way there?

**Comment by loebl** [comment]
I encountered the same issue. As said before, the 3ds importer generated smooth normals for meshes without smoothing groups. This created some shading artifacts. The attached patch changes the code to only generate smooth normals for faces with smoothing groups assigned. We are currently testing if this is acceptable for our models.

If smooth normals for the whole model are desired, using `aiProcess_GenSmoothNormals` in post-processing yields better/controllable results (by setting `AI_CONFIG_PP_GSN_MAX_SMOOTHING_ANGLE`).

If preferred I can open a pull request with the patch.

```
diff --git a/code/SGSpatialSort.cpp b/code/SGSpatialSort.cpp
index 7a80381e..1b481511 100644
--- a/code/SGSpatialSort.cpp
+++ b/code/SGSpatialSort.cpp
@@ -140,27 +140,20 @@ void SGSpatialSort::FindPositions( const aiVector3D& pPosition,
     }
     else
     {
-        // if the given smoothing group is 0, we'll return all surrounding vertices
-        if (!pSG)
+        // Only return vertices if they have a smoothing group
+        if (pSG)
         {
-            while( it->mDistance < maxDist)
+            while (it->mDistance < maxDist)
             {
-                if((it->mPosition - pPosition).SquareLength() < squareEpsilon)
-                    poResults.push_back( it->mIndex);
+                if ((it->mPosition - pPosition).SquareLength() < squareEpsilon &&
+                    (it->mSmoothGroups & pSG || !it->mSmoothGroups))
+                {
+                    poResults.push_back(it->mIndex);
+                }
                 ++it;
-                if( end == it)break;
+                if (end == it)break;
             }
         }
-        else while( it->mDistance < maxDist)
-        {
-            if((it->mPosition - pPosition).SquareLength() < squareEpsilon &&
-                (it->mSmoothGroups & pSG || !it->mSmoothGroups))
-            {
-                poResults.push_back( it->mIndex);
-            }
-            ++it;
-            if( end == it)break;
-        }
     }
 }
```

**Comment by agerrius** [comment]
Sorry for the late reply, but I finally had the time to test the suggestion above. Unfortunately I my case it doesn't work. But it gives me starting point of where to look, so I'll see if I can find something.

**Comment by agerrius** [comment]
The above suggestion did not work for me, but by changing the file SmoothingGroups.inl I got it working here. I ignored faces with a smoothing group of 0 in there and just copied the old normal over for those. For my test objects that works fine.

Not sure if this is a proper fix, so I'm attaching the modified code so other can test.

[SmoothingGroups.zip](https://github.com/assimp/assimp/files/1729193/SmoothingGroups.zip)
